src/util/json.cpp

```diff
@@ -162,3 +164,49 @@ void jsont::swap(jsont &other)
   other.object.swap(object);
   other.value.swap(value);
 }
+
+bool operator==(const jsont &left, const jsont &right)
+{
+  if(left.kind != right.kind)
+    return false;
+  switch(left.kind)
+  {
+  case jsont::kindt::J_NULL:
+    return true;
+  case jsont::kindt::J_TRUE:
+    return true;
+  case jsont::kindt::J_FALSE:
+    return true;
+  case jsont::kindt::J_NUMBER:
+    return left.value == right.value;
+  case jsont::kindt::J_STRING:
+    return left.value == right.value;
+  case jsont::kindt::J_ARRAY:
+  {
+    const auto &left_array = static_cast<const json_arrayt &>(left);
+    const auto &right_array = static_cast<const json_arrayt &>(right);
+    return left_array.size() == right_array.size() &&
+           std::equal(
+             left_array.begin(), left_array.end(), right_array.begin());
+  }
+  case jsont::kindt::J_OBJECT:
+  {
+    const auto &left_object = static_cast<const json_objectt &>(left);
+    const auto &right_object = static_cast<const json_objectt &>(right);
+    const auto yes = [](const std::pair<std::string, jsont> &) { return true; };
+    const size_t left_size = narrow<size_t>(
+      std::count_if(left_object.begin(), left_object.end(), yes));
```

Comment:
`std::distance(left_object.begin(), left_object.end())`

Reply:
Or since `object` is `std::map`, just use `.size()`. At present since it is `map` you could also use `zip` and safely assume the keys occur in the same order, if guarded by a static-assert checking that `typeof(object)` really is an (ordered) map.

Reply:
`json_objectt` is not a `map`, it just has some `map`-like methods exposed. `size` sadly is not one of them.

Reply:
`json_objectt` is not, but its member `object` is. `zip` lives in https://github.com/diffblue/cbmc/blob/develop/src/util/range.h

Reply:
Sure, but if you guard with an appropriate `static_assert` (ideally there would be some type trait `is_ordered`) then you could be sure this code will be changed as/when we do that

Reply:
`object` is a protected property.

There isn't an `is_ordered` trait. Being "ordered" is a runtime property.

If I didn't have to worry about future proofing - `zip` is unnecessary - `std::equal(b1, e1, b2)` is the tool for this job.

Reply:
At least expose a `size` property rather on `json_objectt` rather than hacking around it with the existing interface. The ordered vs. unordered thing I'll pass on.